<!DOCTYPE html>
<!-- saved from url=(0058)https://satijalab.org/seurat/articles/pbmc3k_tutorial.html -->
<html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8">


<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Seurat - Guided Clustering Tutorial • Seurat</title>
<!-- jquery --><script type="text/javascript" async="" src="./Seurat - Guided Clustering Tutorial • Seurat_files/analytics.js"></script><script src="./Seurat - Guided Clustering Tutorial • Seurat_files/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="./Seurat - Guided Clustering Tutorial • Seurat_files/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="./Seurat - Guided Clustering Tutorial • Seurat_files/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="./Seurat - Guided Clustering Tutorial • Seurat_files/bootstrap-toc.css">
<script src="./Seurat - Guided Clustering Tutorial • Seurat_files/bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="./Seurat - Guided Clustering Tutorial • Seurat_files/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="./Seurat - Guided Clustering Tutorial • Seurat_files/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="./Seurat - Guided Clustering Tutorial • Seurat_files/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="./Seurat - Guided Clustering Tutorial • Seurat_files/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="./Seurat - Guided Clustering Tutorial • Seurat_files/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="./Seurat - Guided Clustering Tutorial • Seurat_files/pkgdown.css" rel="stylesheet">
<script src="./Seurat - Guided Clustering Tutorial • Seurat_files/pkgdown.js"></script><meta property="og:title" content="Seurat - Guided Clustering Tutorial">
<meta property="og:description" content="Seurat">
<!-- mathjax --><script src="./Seurat - Guided Clustering Tutorial • Seurat_files/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><style type="text/css">.MathJax_Preview {color: #888}
#MathJax_Message {position: fixed; left: 1em; bottom: 1.5em; background-color: #E6E6E6; border: 1px solid #959595; margin: 0px; padding: 2px 8px; z-index: 102; color: black; font-size: 80%; width: auto; white-space: nowrap}
#MathJax_MSIE_Frame {position: absolute; top: 0; left: 0; width: 0px; z-index: 101; border: 0px; margin: 0px; padding: 0px}
.MathJax_Error {color: #CC0000; font-style: italic}
</style><script src="./Seurat - Guided Clustering Tutorial • Seurat_files/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><style type="text/css">.MathJax_Hover_Frame {border-radius: .25em; -webkit-border-radius: .25em; -moz-border-radius: .25em; -khtml-border-radius: .25em; box-shadow: 0px 0px 15px #83A; -webkit-box-shadow: 0px 0px 15px #83A; -moz-box-shadow: 0px 0px 15px #83A; -khtml-box-shadow: 0px 0px 15px #83A; border: 1px solid #A6D ! important; display: inline-block; position: absolute}
.MathJax_Menu_Button .MathJax_Hover_Arrow {position: absolute; cursor: pointer; display: inline-block; border: 2px solid #AAA; border-radius: 4px; -webkit-border-radius: 4px; -moz-border-radius: 4px; -khtml-border-radius: 4px; font-family: 'Courier New',Courier; font-size: 9px; color: #F0F0F0}
.MathJax_Menu_Button .MathJax_Hover_Arrow span {display: block; background-color: #AAA; border: 1px solid; border-radius: 3px; line-height: 0; padding: 4px}
.MathJax_Hover_Arrow:hover {color: white!important; border: 2px solid #CCC!important}
.MathJax_Hover_Arrow:hover span {background-color: #CCC!important}
</style><style type="text/css">#MathJax_Zoom {position: absolute; background-color: #F0F0F0; overflow: auto; display: block; z-index: 301; padding: .5em; border: 1px solid black; margin: 0; font-weight: normal; font-style: normal; text-align: left; text-indent: 0; text-transform: none; line-height: normal; letter-spacing: normal; word-spacing: normal; word-wrap: normal; white-space: nowrap; float: none; -webkit-box-sizing: content-box; -moz-box-sizing: content-box; box-sizing: content-box; box-shadow: 5px 5px 15px #AAAAAA; -webkit-box-shadow: 5px 5px 15px #AAAAAA; -moz-box-shadow: 5px 5px 15px #AAAAAA; -khtml-box-shadow: 5px 5px 15px #AAAAAA; filter: progid:DXImageTransform.Microsoft.dropshadow(OffX=2, OffY=2, Color='gray', Positive='true')}
#MathJax_ZoomOverlay {position: absolute; left: 0; top: 0; z-index: 300; display: inline-block; width: 100%; height: 100%; border: 0; padding: 0; margin: 0; background-color: white; opacity: 0; filter: alpha(opacity=0)}
#MathJax_ZoomFrame {position: relative; display: inline-block; height: 0; width: 0}
#MathJax_ZoomEventTrap {position: absolute; left: 0; top: 0; z-index: 302; display: inline-block; border: 0; padding: 0; margin: 0; background-color: white; opacity: 0; filter: alpha(opacity=0)}
</style><style type="text/css">#MathJax_About {position: fixed; left: 50%; width: auto; text-align: center; border: 3px outset; padding: 1em 2em; background-color: #DDDDDD; color: black; cursor: default; font-family: message-box; font-size: 120%; font-style: normal; text-indent: 0; text-transform: none; line-height: normal; letter-spacing: normal; word-spacing: normal; word-wrap: normal; white-space: nowrap; float: none; z-index: 201; border-radius: 15px; -webkit-border-radius: 15px; -moz-border-radius: 15px; -khtml-border-radius: 15px; box-shadow: 0px 10px 20px #808080; -webkit-box-shadow: 0px 10px 20px #808080; -moz-box-shadow: 0px 10px 20px #808080; -khtml-box-shadow: 0px 10px 20px #808080; filter: progid:DXImageTransform.Microsoft.dropshadow(OffX=2, OffY=2, Color='gray', Positive='true')}
#MathJax_About.MathJax_MousePost {outline: none}
.MathJax_Menu {position: absolute; background-color: white; color: black; width: auto; padding: 5px 0px; border: 1px solid #CCCCCC; margin: 0; cursor: default; font: menu; text-align: left; text-indent: 0; text-transform: none; line-height: normal; letter-spacing: normal; word-spacing: normal; word-wrap: normal; white-space: nowrap; float: none; z-index: 201; border-radius: 5px; -webkit-border-radius: 5px; -moz-border-radius: 5px; -khtml-border-radius: 5px; box-shadow: 0px 10px 20px #808080; -webkit-box-shadow: 0px 10px 20px #808080; -moz-box-shadow: 0px 10px 20px #808080; -khtml-box-shadow: 0px 10px 20px #808080; filter: progid:DXImageTransform.Microsoft.dropshadow(OffX=2, OffY=2, Color='gray', Positive='true')}
.MathJax_MenuItem {padding: 1px 2em; background: transparent}
.MathJax_MenuArrow {position: absolute; right: .5em; padding-top: .25em; color: #666666; font-size: .75em}
.MathJax_MenuActive .MathJax_MenuArrow {color: white}
.MathJax_MenuArrow.RTL {left: .5em; right: auto}
.MathJax_MenuCheck {position: absolute; left: .7em}
.MathJax_MenuCheck.RTL {right: .7em; left: auto}
.MathJax_MenuRadioCheck {position: absolute; left: .7em}
.MathJax_MenuRadioCheck.RTL {right: .7em; left: auto}
.MathJax_MenuLabel {padding: 1px 2em 3px 1.33em; font-style: italic}
.MathJax_MenuRule {border-top: 1px solid #DDDDDD; margin: 4px 3px}
.MathJax_MenuDisabled {color: GrayText}
.MathJax_MenuActive {background-color: #606872; color: white}
.MathJax_MenuDisabled:focus, .MathJax_MenuLabel:focus {background-color: #E8E8E8}
.MathJax_ContextMenu:focus {outline: none}
.MathJax_ContextMenu .MathJax_MenuItem:focus {outline: none}
#MathJax_AboutClose {top: .2em; right: .2em}
.MathJax_Menu .MathJax_MenuClose {top: -10px; left: -10px}
.MathJax_MenuClose {position: absolute; cursor: pointer; display: inline-block; border: 2px solid #AAA; border-radius: 18px; -webkit-border-radius: 18px; -moz-border-radius: 18px; -khtml-border-radius: 18px; font-family: 'Courier New',Courier; font-size: 24px; color: #F0F0F0}
.MathJax_MenuClose span {display: block; background-color: #AAA; border: 1.5px solid; border-radius: 18px; -webkit-border-radius: 18px; -moz-border-radius: 18px; -khtml-border-radius: 18px; line-height: 0; padding: 8px 0 6px}
.MathJax_MenuClose:hover {color: white!important; border: 2px solid #CCC!important}
.MathJax_MenuClose:hover span {background-color: #CCC!important}
.MathJax_MenuClose:hover:focus {outline: none}
</style><style type="text/css">.MathJax_Preview .MJXf-math {color: inherit!important}
</style><style type="text/css">.MJX_Assistive_MathML {position: absolute!important; top: 0; left: 0; clip: rect(1px, 1px, 1px, 1px); padding: 1px 0 0 0!important; border: 0!important; height: 1px!important; width: 1px!important; overflow: hidden!important; display: block!important; -webkit-touch-callout: none; -webkit-user-select: none; -khtml-user-select: none; -moz-user-select: none; -ms-user-select: none; user-select: none}
.MJX_Assistive_MathML.MJX_Assistive_MathML_Block {width: 100%!important}
</style><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async="" src="./Seurat - Guided Clustering Tutorial • Seurat_files/js"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-82807227-1');
</script>
<style type="text/css">.MJXp-script {font-size: .8em}
.MJXp-right {-webkit-transform-origin: right; -moz-transform-origin: right; -ms-transform-origin: right; -o-transform-origin: right; transform-origin: right}
.MJXp-bold {font-weight: bold}
.MJXp-italic {font-style: italic}
.MJXp-scr {font-family: MathJax_Script,'Times New Roman',Times,STIXGeneral,serif}
.MJXp-frak {font-family: MathJax_Fraktur,'Times New Roman',Times,STIXGeneral,serif}
.MJXp-sf {font-family: MathJax_SansSerif,'Times New Roman',Times,STIXGeneral,serif}
.MJXp-cal {font-family: MathJax_Caligraphic,'Times New Roman',Times,STIXGeneral,serif}
.MJXp-mono {font-family: MathJax_Typewriter,'Times New Roman',Times,STIXGeneral,serif}
.MJXp-largeop {font-size: 150%}
.MJXp-largeop.MJXp-int {vertical-align: -.2em}
.MJXp-math {display: inline-block; line-height: 1.2; text-indent: 0; font-family: 'Times New Roman',Times,STIXGeneral,serif; white-space: nowrap; border-collapse: collapse}
.MJXp-display {display: block; text-align: center; margin: 1em 0}
.MJXp-math span {display: inline-block}
.MJXp-box {display: block!important; text-align: center}
.MJXp-box:after {content: " "}
.MJXp-rule {display: block!important; margin-top: .1em}
.MJXp-char {display: block!important}
.MJXp-mo {margin: 0 .15em}
.MJXp-mfrac {margin: 0 .125em; vertical-align: .25em}
.MJXp-denom {display: inline-table!important; width: 100%}
.MJXp-denom > * {display: table-row!important}
.MJXp-surd {vertical-align: top}
.MJXp-surd > * {display: block!important}
.MJXp-script-box > *  {display: table!important; height: 50%}
.MJXp-script-box > * > * {display: table-cell!important; vertical-align: top}
.MJXp-script-box > *:last-child > * {vertical-align: bottom}
.MJXp-script-box > * > * > * {display: block!important}
.MJXp-mphantom {visibility: hidden}
.MJXp-munderover, .MJXp-munder {display: inline-table!important}
.MJXp-over {display: inline-block!important; text-align: center}
.MJXp-over > * {display: block!important}
.MJXp-munderover > *, .MJXp-munder > * {display: table-row!important}
.MJXp-mtable {vertical-align: .25em; margin: 0 .125em}
.MJXp-mtable > * {display: inline-table!important; vertical-align: middle}
.MJXp-mtr {display: table-row!important}
.MJXp-mtd {display: table-cell!important; text-align: center; padding: .5em 0 0 .5em}
.MJXp-mtr > .MJXp-mtd:first-child {padding-left: 0}
.MJXp-mtr:first-child > .MJXp-mtd {padding-top: 0}
.MJXp-mlabeledtr {display: table-row!important}
.MJXp-mlabeledtr > .MJXp-mtd:first-child {padding-left: 0}
.MJXp-mlabeledtr:first-child > .MJXp-mtd {padding-top: 0}
.MJXp-merror {background-color: #FFFF88; color: #CC0000; border: 1px solid #CC0000; padding: 1px 3px; font-style: normal; font-size: 90%}
.MJXp-scale0 {-webkit-transform: scaleX(.0); -moz-transform: scaleX(.0); -ms-transform: scaleX(.0); -o-transform: scaleX(.0); transform: scaleX(.0)}
.MJXp-scale1 {-webkit-transform: scaleX(.1); -moz-transform: scaleX(.1); -ms-transform: scaleX(.1); -o-transform: scaleX(.1); transform: scaleX(.1)}
.MJXp-scale2 {-webkit-transform: scaleX(.2); -moz-transform: scaleX(.2); -ms-transform: scaleX(.2); -o-transform: scaleX(.2); transform: scaleX(.2)}
.MJXp-scale3 {-webkit-transform: scaleX(.3); -moz-transform: scaleX(.3); -ms-transform: scaleX(.3); -o-transform: scaleX(.3); transform: scaleX(.3)}
.MJXp-scale4 {-webkit-transform: scaleX(.4); -moz-transform: scaleX(.4); -ms-transform: scaleX(.4); -o-transform: scaleX(.4); transform: scaleX(.4)}
.MJXp-scale5 {-webkit-transform: scaleX(.5); -moz-transform: scaleX(.5); -ms-transform: scaleX(.5); -o-transform: scaleX(.5); transform: scaleX(.5)}
.MJXp-scale6 {-webkit-transform: scaleX(.6); -moz-transform: scaleX(.6); -ms-transform: scaleX(.6); -o-transform: scaleX(.6); transform: scaleX(.6)}
.MJXp-scale7 {-webkit-transform: scaleX(.7); -moz-transform: scaleX(.7); -ms-transform: scaleX(.7); -o-transform: scaleX(.7); transform: scaleX(.7)}
.MJXp-scale8 {-webkit-transform: scaleX(.8); -moz-transform: scaleX(.8); -ms-transform: scaleX(.8); -o-transform: scaleX(.8); transform: scaleX(.8)}
.MJXp-scale9 {-webkit-transform: scaleX(.9); -moz-transform: scaleX(.9); -ms-transform: scaleX(.9); -o-transform: scaleX(.9); transform: scaleX(.9)}
.MathJax_PHTML .noError {vertical-align: ; font-size: 90%; text-align: left; color: black; padding: 1px 3px; border: 1px solid}
</style></head>
<body data-spy="scroll" data-target="#toc" style="padding-top: 70px;"><div id="MathJax_Message" style="display: none;"></div>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top headroom headroom--top headroom--not-bottom" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="https://satijalab.org/seurat/index.html">Seurat</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="" data-original-title="Released version">4.0.6</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="https://satijalab.org/seurat/articles/install.html">Install</a>
</li>
<li>
  <a href="https://satijalab.org/seurat/articles/get_started.html">Get started</a>
</li>
<li class="dropdown active">
  <a href="https://satijalab.org/seurat/articles/pbmc3k_tutorial.html#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-header">Introductory Vignettes</li>
    <li class="active">
      <a href="https://satijalab.org/seurat/articles/pbmc3k_tutorial.html">PBMC 3K guided tutorial</a>
    </li>
    <li>
      <a href="https://satijalab.org/seurat/articles/multimodal_vignette.html">Using Seurat with multi-modal data</a>
    </li>
    <li>
      <a href="https://satijalab.org/seurat/articles/spatial_vignette.html">Analysis, visualization, and integration of spatial datasets with Seurat</a>
    </li>
    <li class="divider">
    </li><li class="dropdown-header">Data Integration</li>
    <li>
      <a href="https://satijalab.org/seurat/articles/integration_introduction.html">Introduction to scRNA-seq integration</a>
    </li>
    <li>
      <a href="https://satijalab.org/seurat/articles/integration_mapping.html">Mapping and annotating query datasets</a>
    </li>
    <li>
      <a href="https://satijalab.org/seurat/articles/integration_rpca.html">Fast integration using reciprocal PCA (RPCA)</a>
    </li>
    <li>
      <a href="https://satijalab.org/seurat/articles/integration_large_datasets.html">Tips for integrating large datasets</a>
    </li>
    <li>
      <a href="https://satijalab.org/seurat/articles/atacseq_integration_vignette.html">Integrating scRNA-seq and scATAC-seq data</a>
    </li>
    <li>
      <a href="https://satijalab.org/seurat/articles/multimodal_reference_mapping.html">Multimodal reference mapping</a>
    </li>
    <li class="divider">
    </li><li class="dropdown-header">New Statistical Methods</li>
    <li>
      <a href="https://satijalab.org/seurat/articles/weighted_nearest_neighbor_analysis.html">Weighted Nearest Neighbor Analysis</a>
    </li>
    <li>
      <a href="https://satijalab.org/seurat/articles/mixscape_vignette.html">Mixscape Vignette</a>
    </li>
    <li>
      <a href="https://satijalab.org/seurat/articles/sctransform_vignette.html">Using sctransform in Seurat</a>
    </li>
    <li>
      <a href="https://satijalab.org/seurat/articles/sctransform_v2_vignette.html">SCTransform, v2 regularization</a>
    </li>
    <li class="divider">
    </li><li class="dropdown-header">Other</li>
    <li>
      <a href="https://satijalab.org/seurat/articles/visualization_vignette.html">Data visualization vignette</a>
    </li>
    <li>
      <a href="https://satijalab.org/seurat/articles/cell_cycle_vignette.html">Cell-cycle scoring and regression</a>
    </li>
    <li>
      <a href="https://satijalab.org/seurat/articles/de_vignette.html">Differential expression testing</a>
    </li>
    <li>
      <a href="https://satijalab.org/seurat/articles/hashing_vignette.html">Demultiplexing with hashtag oligos (HTOs)</a>
    </li>
    <li>
      <a href="https://satijalab.org/seurat/articles/conversion_vignette.html">Interoperability between single-cell object formats</a>
    </li>
    <li>
      <a href="https://satijalab.org/seurat/articles/future_vignette.html">Parallelization in Seurat with future</a>
    </li>
    <li>
      <a href="https://satijalab.org/seurat/articles/dim_reduction_vignette.html">Dimensional reduction vignette</a>
    </li>
    <li>
      <a href="https://satijalab.org/seurat/articles/essential_commands.html">Seurat essential commands list</a>
    </li>
    <li>
      <a href="https://satijalab.org/seurat/articles/interaction_vignette.html">Seurat interaction tips</a>
    </li>
    <li>
      <a href="https://satijalab.org/seurat/articles/merge_vignette.html">Merging Seurat objects</a>
    </li>
  </ul>
</li>
<li>
  <a href="https://satijalab.org/seurat/articles/extensions.html">Extensions</a>
</li>
<li>
  <a href="https://github.com/satijalab/seurat/discussions">FAQ</a>
</li>
<li>
  <a href="https://satijalab.org/seurat/news/index.html">News</a>
</li>
<li>
  <a href="https://satijalab.org/seurat/reference/index.html">Reference</a>
</li>
<li>
  <a href="https://satijalab.org/seurat/articles/archive.html">Archive</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://satijalab.org/">
    <span class="fa fa-home"></span>
     
  </a>
</li>
<li>
  <a href="https://github.com/satijalab/seurat">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="./Seurat - Guided Clustering Tutorial • Seurat_files/header-attrs.js"></script><script src="./Seurat - Guided Clustering Tutorial • Seurat_files/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip="">Seurat - Guided Clustering Tutorial</h1>
            
            <h4 class="date">Compiled: January 11, 2022</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/satijalab/seurat/blob/master/vignettes/pbmc3k_tutorial.Rmd"><code>vignettes/pbmc3k_tutorial.Rmd</code></a></small>
      <div class="hidden name"><code>pbmc3k_tutorial.Rmd</code></div>

    </div>

    
    
<hr>
<div id="setup-the-seurat-object" class="section level1">
<h1 id="setup-the-seurat-object-1">
<a href="https://satijalab.org/seurat/articles/pbmc3k_tutorial.html#setup-the-seurat-object" class="anchor"></a>Setup the Seurat Object</h1>
<p>For this tutorial, we will be analyzing the a dataset of Peripheral Blood Mononuclear Cells (PBMC) freely available from 10X Genomics. There are 2,700 single cells that were sequenced on the Illumina NextSeq 500. The raw data can be found <a href="https://cf.10xgenomics.com/samples/cell/pbmc3k/pbmc3k_filtered_gene_bc_matrices.tar.gz">here</a>.</p>
<p>We start by reading in the data. The <code><a href="https://satijalab.org/seurat/reference/Read10X.html">Read10X()</a></code> function reads in the output of the <a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/what-is-cell-ranger">cellranger</a> pipeline from 10X, returning a unique molecular identified (UMI) count matrix. The values in this matrix represent the number of molecules for each feature (i.e.&nbsp;gene; row) that are detected in each cell (column).</p>
<p>We next use the count matrix to create a <code>Seurat</code> object. The object serves as a container that contains both data (like the count matrix) and analysis (like PCA, or clustering results) for a single-cell dataset. For a technical discussion of the <code>Seurat</code> object structure, check out our <a href="https://github.com/satijalab/seurat/wiki">GitHub Wiki</a>. For example, the count matrix is stored in <code>pbmc[["RNA"]]@counts</code>.</p>
<div class="sourceCode hasCopyButton" id="cb1"><button type="button" class="btn btn-primary btn-copy-ex" title="" aria-label="Copy to clipboard" data-toggle="tooltip" data-placement="left auto" data-trigger="hover" data-clipboard-copy="" data-original-title="Copy to clipboard"><i class="fa fa-copy"></i></button><pre class="downlit sourceCode r"><code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org/">dplyr</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat">Seurat</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com/">patchwork</a></span><span class="op">)</span>

<span class="co"># Load the PBMC dataset</span>
<span class="va">pbmc.data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/Read10X.html">Read10X</a></span><span class="op">(</span>data.dir <span class="op">=</span> <span class="st">"../data/pbmc3k/filtered_gene_bc_matrices/hg19/"</span><span class="op">)</span>
<span class="co"># Initialize the Seurat object with the raw (non-normalized data).</span>
<span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/CreateSeuratObject.html">CreateSeuratObject</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="va">pbmc.data</span>, project <span class="op">=</span> <span class="st">"pbmc3k"</span>, min.cells <span class="op">=</span> <span class="fl">3</span>, min.features <span class="op">=</span> <span class="fl">200</span><span class="op">)</span>
<span class="va">pbmc</span></code></pre></div>
<pre><code>## An object of class Seurat 
## 13714 features across 2700 samples within 1 assay 
## Active assay: RNA (13714 features, 0 variable features)</code></pre>
<details><p></p><summary><strong>What does data in a count matrix look like?</strong></summary><p></p>
<div class="sourceCode hasCopyButton" id="cb3"><button type="button" class="btn btn-primary btn-copy-ex" title="" aria-label="Copy to clipboard" data-toggle="tooltip" data-placement="left auto" data-trigger="hover" data-clipboard-copy="" data-original-title="Copy to clipboard"><i class="fa fa-copy"></i></button><pre class="downlit sourceCode r"><code class="sourceCode R"><span class="co"># Lets examine a few genes in the first thirty cells</span>
<span class="va">pbmc.data</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"CD3D"</span>, <span class="st">"TCL1A"</span>, <span class="st">"MS4A1"</span><span class="op">)</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">30</span><span class="op">]</span></code></pre></div>
<pre><code>## 3 x 30 sparse Matrix of class "dgCMatrix"
##                                                                    
## CD3D  4 . 10 . . 1 2 3 1 . . 2 7 1 . . 1 3 . 2  3 . . . . . 3 4 1 5
## TCL1A . .  . . . . . . 1 . . . . . . . . . . .  . 1 . . . . . . . .
## MS4A1 . 6  . . . . . . 1 1 1 . . . . . . . . . 36 1 2 . . 2 . . . .</code></pre>
<p>The <code>.</code> values in the matrix represent 0s (no molecules detected). Since most values in an scRNA-seq matrix are 0, Seurat uses a sparse-matrix representation whenever possible. This results in significant memory and speed savings for Drop-seq/inDrop/10x data.</p>
<div class="sourceCode hasCopyButton" id="cb5"><button type="button" class="btn btn-primary btn-copy-ex" title="" aria-label="Copy to clipboard" data-toggle="tooltip" data-placement="left auto" data-trigger="hover" data-clipboard-copy="" data-original-title="Copy to clipboard"><i class="fa fa-copy"></i></button><pre class="downlit sourceCode r"><code class="sourceCode R"><span class="va">dense.size</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/object.size.html">object.size</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span><span class="op">(</span><span class="va">pbmc.data</span><span class="op">)</span><span class="op">)</span>
<span class="va">dense.size</span></code></pre></div>
<pre><code>## 709591472 bytes</code></pre>
<div class="sourceCode hasCopyButton" id="cb7"><button type="button" class="btn btn-primary btn-copy-ex" title="" aria-label="Copy to clipboard" data-toggle="tooltip" data-placement="left auto" data-trigger="hover" data-clipboard-copy="" data-original-title="Copy to clipboard"><i class="fa fa-copy"></i></button><pre class="downlit sourceCode r"><code class="sourceCode R"><span class="va">sparse.size</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/object.size.html">object.size</a></span><span class="op">(</span><span class="va">pbmc.data</span><span class="op">)</span>
<span class="va">sparse.size</span></code></pre></div>
<pre><code>## 29905192 bytes</code></pre>
<div class="sourceCode hasCopyButton" id="cb9"><button type="button" class="btn btn-primary btn-copy-ex" title="" aria-label="Copy to clipboard" data-toggle="tooltip" data-placement="left auto" data-trigger="hover" data-clipboard-copy="" data-original-title="Copy to clipboard"><i class="fa fa-copy"></i></button><pre class="downlit sourceCode r"><code class="sourceCode R"><span class="va">dense.size</span><span class="op">/</span><span class="va">sparse.size</span></code></pre></div>
<pre><code>## 23.7 bytes</code></pre>
</details><p>&nbsp;</p>
</div>
<div id="standard-pre-processing-workflow" class="section level1">
<h1 id="standard-pre-processing-workflow-1">
<a href="https://satijalab.org/seurat/articles/pbmc3k_tutorial.html#standard-pre-processing-workflow" class="anchor"></a>Standard pre-processing workflow</h1>
<p>The steps below encompass the standard pre-processing workflow for scRNA-seq data in Seurat. These represent the selection and filtration of cells based on QC metrics, data normalization and scaling, and the detection of highly variable features.</p>
<div id="qc-and-selecting-cells-for-further-analysis" class="section level2">
<h2 id="qc-and-selecting-cells-for-further-analysis-1">
<a href="https://satijalab.org/seurat/articles/pbmc3k_tutorial.html#qc-and-selecting-cells-for-further-analysis" class="anchor"></a>QC and selecting cells for further analysis</h2>
<p>Seurat allows you to easily explore QC metrics and filter cells based on any user-defined criteria. A few QC metrics <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4758103/">commonly used</a> by the community include</p>
<ul>
<li>The number of unique genes detected in each cell.
<ul>
<li>Low-quality cells or empty droplets will often have very few genes</li>
<li>Cell doublets or multiplets may exhibit an aberrantly high gene count</li>
</ul>
</li>
<li>Similarly, the total number of molecules detected within a cell (correlates strongly with unique genes)</li>
<li>The percentage of reads that map to the mitochondrial genome
<ul>
<li>Low-quality / dying cells often exhibit extensive mitochondrial contamination</li>
<li>We calculate mitochondrial QC metrics with the <code><a href="https://satijalab.org/seurat/reference/PercentageFeatureSet.html">PercentageFeatureSet()</a></code> function, which calculates the percentage of counts originating from a set of features</li>
<li>We use the set of all genes starting with <code>MT-</code> as a set of mitochondrial genes</li>
</ul>
</li>
</ul>
<div class="sourceCode hasCopyButton" id="cb11"><button type="button" class="btn btn-primary btn-copy-ex" title="" aria-label="Copy to clipboard" data-toggle="tooltip" data-placement="left auto" data-trigger="hover" data-clipboard-copy="" data-original-title="Copy to clipboard"><i class="fa fa-copy"></i></button><pre class="downlit sourceCode r"><code class="sourceCode R"><span class="co"># The [[ operator can add columns to object metadata. This is a great place to stash QC stats</span>
<span class="va">pbmc</span><span class="op">[[</span><span class="st">"percent.mt"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/PercentageFeatureSet.html">PercentageFeatureSet</a></span><span class="op">(</span><span class="va">pbmc</span>, pattern <span class="op">=</span> <span class="st">"^MT-"</span><span class="op">)</span></code></pre></div>
<details><p></p><summary><strong>Where are QC metrics stored in Seurat?</strong></summary><p></p>
<ul>
<li>The number of unique genes and total molecules are automatically calculated during <code><a href="https://mojaveazure.github.io/seurat-object/reference/CreateSeuratObject.html">CreateSeuratObject()</a></code>
<ul>
<li>You can find them stored in the object meta data</li>
</ul>
</li>
</ul>
<div class="sourceCode hasCopyButton" id="cb12"><button type="button" class="btn btn-primary btn-copy-ex" title="" aria-label="Copy to clipboard" data-toggle="tooltip" data-placement="left auto" data-trigger="hover" data-clipboard-copy="" data-original-title="Copy to clipboard"><i class="fa fa-copy"></i></button><pre class="downlit sourceCode r"><code class="sourceCode R"><span class="co"># Show QC metrics for the first 5 cells</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">@</span><span class="va">meta.data</span>, <span class="fl">5</span><span class="op">)</span></code></pre></div>
<pre><code>##                  orig.ident nCount_RNA nFeature_RNA percent.mt
## AAACATACAACCAC-1     pbmc3k       2419          779  3.0177759
## AAACATTGAGCTAC-1     pbmc3k       4903         1352  3.7935958
## AAACATTGATCAGC-1     pbmc3k       3147         1129  0.8897363
## AAACCGTGCTTCCG-1     pbmc3k       2639          960  1.7430845
## AAACCGTGTATGCG-1     pbmc3k        980          521  1.2244898</code></pre>
</details><p>&nbsp;</p>
<p>In the example below, we visualize QC metrics, and use these to filter cells.</p>
<ul>
<li>We filter cells that have unique feature counts over 2,500 or less than 200</li>
<li>We filter cells that have &gt;5% mitochondrial counts</li>
</ul>
<div class="sourceCode hasCopyButton" id="cb14"><button type="button" class="btn btn-primary btn-copy-ex" title="" aria-label="Copy to clipboard" data-toggle="tooltip" data-placement="left auto" data-trigger="hover" data-clipboard-copy="" data-original-title="Copy to clipboard"><i class="fa fa-copy"></i></button><pre class="downlit sourceCode r"><code class="sourceCode R"><span class="co"># Visualize QC metrics as a violin plot</span>
<span class="fu"><a href="https://satijalab.org/seurat/reference/VlnPlot.html">VlnPlot</a></span><span class="op">(</span><span class="va">pbmc</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"nFeature_RNA"</span>, <span class="st">"nCount_RNA"</span>, <span class="st">"percent.mt"</span><span class="op">)</span>, ncol <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></code></pre></div>
<p><img src="./Seurat - Guided Clustering Tutorial • Seurat_files/qc2-1.png" width="1248"></p>
<div class="sourceCode hasCopyButton" id="cb15"><button type="button" class="btn btn-primary btn-copy-ex" title="" aria-label="Copy to clipboard" data-toggle="tooltip" data-placement="left auto" data-trigger="hover" data-clipboard-copy="" data-original-title="Copy to clipboard"><i class="fa fa-copy"></i></button><pre class="downlit sourceCode r"><code class="sourceCode R"><span class="co"># FeatureScatter is typically used to visualize feature-feature relationships, but can be used</span>
<span class="co"># for anything calculated by the object, i.e. columns in object metadata, PC scores etc.</span>

<span class="va">plot1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FeatureScatter.html">FeatureScatter</a></span><span class="op">(</span><span class="va">pbmc</span>, feature1 <span class="op">=</span> <span class="st">"nCount_RNA"</span>, feature2 <span class="op">=</span> <span class="st">"percent.mt"</span><span class="op">)</span>
<span class="va">plot2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FeatureScatter.html">FeatureScatter</a></span><span class="op">(</span><span class="va">pbmc</span>, feature1 <span class="op">=</span> <span class="st">"nCount_RNA"</span>, feature2 <span class="op">=</span> <span class="st">"nFeature_RNA"</span><span class="op">)</span>
<span class="va">plot1</span> <span class="op">+</span> <span class="va">plot2</span></code></pre></div>
<p><img src="./Seurat - Guided Clustering Tutorial • Seurat_files/qc2-2.png" width="1248"></p>
<div class="sourceCode hasCopyButton" id="cb16"><button type="button" class="btn btn-primary btn-copy-ex" title="" aria-label="Copy to clipboard" data-toggle="tooltip" data-placement="left auto" data-trigger="hover" data-clipboard-copy="" data-original-title="Copy to clipboard"><i class="fa fa-copy"></i></button><pre class="downlit sourceCode r"><code class="sourceCode R"><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html">subset</a></span><span class="op">(</span><span class="va">pbmc</span>, subset <span class="op">=</span> <span class="va">nFeature_RNA</span> <span class="op">&gt;</span> <span class="fl">200</span> <span class="op">&amp;</span> <span class="va">nFeature_RNA</span> <span class="op">&lt;</span> <span class="fl">2500</span> <span class="op">&amp;</span> <span class="va">percent.mt</span> <span class="op">&lt;</span> <span class="fl">5</span><span class="op">)</span></code></pre></div>
<hr>
</div>
</div>
<div id="normalizing-the-data" class="section level1">
<h1 id="normalizing-the-data-1">
<a href="https://satijalab.org/seurat/articles/pbmc3k_tutorial.html#normalizing-the-data" class="anchor"></a>Normalizing the data</h1>
<p>After removing unwanted cells from the dataset, the next step is to normalize the data. By default, we employ a global-scaling normalization method “LogNormalize” that normalizes the feature expression measurements for each cell by the total expression, multiplies this by a scale factor (10,000 by default), and log-transforms the result. Normalized values are stored in <code>pbmc[["RNA"]]@data</code>.</p>
<div class="sourceCode hasCopyButton" id="cb17"><button type="button" class="btn btn-primary btn-copy-ex" title="" aria-label="Copy to clipboard" data-toggle="tooltip" data-placement="left auto" data-trigger="hover" data-clipboard-copy="" data-original-title="Copy to clipboard"><i class="fa fa-copy"></i></button><pre class="downlit sourceCode r"><code class="sourceCode R"><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/NormalizeData.html">NormalizeData</a></span><span class="op">(</span><span class="va">pbmc</span>, normalization.method <span class="op">=</span> <span class="st">"LogNormalize"</span>, scale.factor <span class="op">=</span> <span class="fl">10000</span><span class="op">)</span></code></pre></div>
<p>For clarity, in this previous line of code (and in future commands), we provide the default values for certain parameters in the function call. However, this isn’t required and the same behavior can be achieved with:</p>
<div class="sourceCode hasCopyButton" id="cb18"><button type="button" class="btn btn-primary btn-copy-ex" title="" aria-label="Copy to clipboard" data-toggle="tooltip" data-placement="left auto" data-trigger="hover" data-clipboard-copy="" data-original-title="Copy to clipboard"><i class="fa fa-copy"></i></button><pre class="downlit sourceCode r"><code class="sourceCode R"><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/NormalizeData.html">NormalizeData</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span></code></pre></div>
</div>
<div id="identification-of-highly-variable-features-feature-selection" class="section level1">
<h1 id="identification-of-highly-variable-features-feature-selection-">
<a href="https://satijalab.org/seurat/articles/pbmc3k_tutorial.html#identification-of-highly-variable-features-feature-selection" class="anchor"></a>Identification of highly variable features (feature selection)</h1>
<p>We next calculate a subset of features that exhibit high cell-to-cell variation in the dataset (i.e, they are highly expressed in some cells, and lowly expressed in others). We and <a href="https://www.nature.com/articles/nmeth.2645">others</a> have found that focusing on these genes in downstream analysis helps to highlight biological signal in single-cell datasets.</p>
<p>Our procedure in Seurat is described in detail <a href="https://doi.org/10.1016/j.cell.2019.05.031">here</a>, and improves on previous versions by directly modeling the mean-variance relationship inherent in single-cell data, and is implemented in the <code><a href="https://satijalab.org/seurat/reference/FindVariableFeatures.html">FindVariableFeatures()</a></code> function. By default, we return 2,000 features per dataset. These will be used in downstream analysis, like PCA.</p>
<div class="sourceCode hasCopyButton" id="cb19"><button type="button" class="btn btn-primary btn-copy-ex" title="" aria-label="Copy to clipboard" data-toggle="tooltip" data-placement="left auto" data-trigger="hover" data-clipboard-copy="" data-original-title="Copy to clipboard"><i class="fa fa-copy"></i></button><pre class="downlit sourceCode r"><code class="sourceCode R"><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindVariableFeatures.html">FindVariableFeatures</a></span><span class="op">(</span><span class="va">pbmc</span>, selection.method <span class="op">=</span> <span class="st">"vst"</span>, nfeatures <span class="op">=</span> <span class="fl">2000</span><span class="op">)</span>

<span class="co"># Identify the 10 most highly variable genes</span>
<span class="va">top10</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/VariableFeatures.html">VariableFeatures</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span>, <span class="fl">10</span><span class="op">)</span>

<span class="co"># plot variable features with and without labels</span>
<span class="va">plot1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/VariableFeaturePlot.html">VariableFeaturePlot</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span>
<span class="va">plot2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/LabelPoints.html">LabelPoints</a></span><span class="op">(</span>plot <span class="op">=</span> <span class="va">plot1</span>, points <span class="op">=</span> <span class="va">top10</span>, repel <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="va">plot1</span> <span class="op">+</span> <span class="va">plot2</span></code></pre></div>
<p><img src="./Seurat - Guided Clustering Tutorial • Seurat_files/var_features-1.png" width="1056"></p>
<hr>
</div>
<div id="scaling-the-data" class="section level1">
<h1 id="scaling-the-data-1">
<a href="https://satijalab.org/seurat/articles/pbmc3k_tutorial.html#scaling-the-data" class="anchor"></a>Scaling the data</h1>
<p>Next, we apply a linear transformation (‘scaling’) that is a standard pre-processing step prior to dimensional reduction techniques like PCA. The <code><a href="https://satijalab.org/seurat/reference/ScaleData.html">ScaleData()</a></code> function:</p>
<ul>
<li>Shifts the expression of each gene, so that the mean expression across cells is 0</li>
<li>Scales the expression of each gene, so that the variance across cells is 1
<ul>
<li>This step gives equal weight in downstream analyses, so that highly-expressed genes do not dominate</li>
</ul>
</li>
<li>The results of this are stored in <code>pbmc[["RNA"]]@scale.data</code>
</li>
</ul>
<div class="sourceCode hasCopyButton" id="cb20"><button type="button" class="btn btn-primary btn-copy-ex" title="" aria-label="Copy to clipboard" data-toggle="tooltip" data-placement="left auto" data-trigger="hover" data-clipboard-copy="" data-original-title="Copy to clipboard"><i class="fa fa-copy"></i></button><pre class="downlit sourceCode r"><code class="sourceCode R"><span class="va">all.genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span>
<span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/ScaleData.html">ScaleData</a></span><span class="op">(</span><span class="va">pbmc</span>, features <span class="op">=</span> <span class="va">all.genes</span><span class="op">)</span></code></pre></div>
<details><p></p><summary><strong>This step takes too long! Can I make it faster?</strong></summary><p></p>
<p>Scaling is an essential step in the Seurat workflow, but only on genes that will be used as input to PCA. Therefore, the default in <code><a href="https://satijalab.org/seurat/reference/ScaleData.html">ScaleData()</a></code> is only to perform scaling on the previously identified variable features (2,000 by default). To do this, omit the <code>features</code> argument in the previous function call, i.e.</p>
<div class="sourceCode hasCopyButton" id="cb21"><button type="button" class="btn btn-primary btn-copy-ex" title="" aria-label="Copy to clipboard" data-toggle="tooltip" data-placement="left auto" data-trigger="hover" data-clipboard-copy="" data-original-title="Copy to clipboard"><i class="fa fa-copy"></i></button><pre class="downlit sourceCode r"><code class="sourceCode R"><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/ScaleData.html">ScaleData</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span></code></pre></div>
Your PCA and clustering results will be unaffected. However, Seurat heatmaps (produced as shown below with <code><a href="https://satijalab.org/seurat/reference/DoHeatmap.html">DoHeatmap()</a></code>) require genes in the heatmap to be scaled, to make sure highly-expressed genes don’t dominate the heatmap. To make sure we don’t leave any genes out of the heatmap later, we are scaling all genes in this tutorial.
</details>
&nbsp;
<details><p></p><summary><strong>How can I remove unwanted sources of variation, as in Seurat v2?</strong></summary><p></p>
<p>In <code>Seurat v2</code> we also use the <code><a href="https://satijalab.org/seurat/reference/ScaleData.html">ScaleData()</a></code> function to remove unwanted sources of variation from a single-cell dataset. For example, we could ‘regress out’ heterogeneity associated with (for example) cell cycle stage, or mitochondrial contamination. These features are still supported in <code><a href="https://satijalab.org/seurat/reference/ScaleData.html">ScaleData()</a></code> in <code>Seurat v3</code>, i.e.:</p>
<div class="sourceCode hasCopyButton" id="cb22"><button type="button" class="btn btn-primary btn-copy-ex" title="" aria-label="Copy to clipboard" data-toggle="tooltip" data-placement="left auto" data-trigger="hover" data-clipboard-copy="" data-original-title="Copy to clipboard"><i class="fa fa-copy"></i></button><pre class="downlit sourceCode r"><code class="sourceCode R"><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/ScaleData.html">ScaleData</a></span><span class="op">(</span><span class="va">pbmc</span>, vars.to.regress <span class="op">=</span> <span class="st">"percent.mt"</span><span class="op">)</span></code></pre></div>
However, particularly for advanced users who would like to use this functionality, we strongly recommend the use of our new normalization workflow, <code><a href="https://satijalab.org/seurat/reference/SCTransform.html">SCTransform()</a></code>. The method is described in our <a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1874-1">paper</a>, with a separate vignette using Seurat v3 <a href="https://satijalab.org/seurat/articles/sctransform_vignette.html">here</a>. As with <code><a href="https://satijalab.org/seurat/reference/ScaleData.html">ScaleData()</a></code>, the function <code><a href="https://satijalab.org/seurat/reference/SCTransform.html">SCTransform()</a></code> also includes a <code>vars.to.regress</code> parameter.
</details><p>&nbsp;</p>
<hr>
</div>
<div id="perform-linear-dimensional-reduction" class="section level1">
<h1 id="perform-linear-dimensional-reduction-1">
<a href="https://satijalab.org/seurat/articles/pbmc3k_tutorial.html#perform-linear-dimensional-reduction" class="anchor"></a>Perform linear dimensional reduction</h1>
<p>Next we perform PCA on the scaled data. By default, only the previously determined variable features are used as input, but can be defined using <code>features</code> argument if you wish to choose a different subset.</p>
<div class="sourceCode hasCopyButton" id="cb23"><button type="button" class="btn btn-primary btn-copy-ex" title="" aria-label="Copy to clipboard" data-toggle="tooltip" data-placement="left auto" data-trigger="hover" data-clipboard-copy="" data-original-title="Copy to clipboard"><i class="fa fa-copy"></i></button><pre class="downlit sourceCode r"><code class="sourceCode R"><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunPCA.html">RunPCA</a></span><span class="op">(</span><span class="va">pbmc</span>, features <span class="op">=</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/VariableFeatures.html">VariableFeatures</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pbmc</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>Seurat provides several useful ways of visualizing both cells and features that define the PCA, including <code>VizDimReduction()</code>, <code><a href="https://satijalab.org/seurat/reference/DimPlot.html">DimPlot()</a></code>, and <code><a href="https://satijalab.org/seurat/reference/DimHeatmap.html">DimHeatmap()</a></code></p>
<div class="sourceCode hasCopyButton" id="cb24"><button type="button" class="btn btn-primary btn-copy-ex" title="" aria-label="Copy to clipboard" data-toggle="tooltip" data-placement="left auto" data-trigger="hover" data-clipboard-copy="" data-original-title="Copy to clipboard"><i class="fa fa-copy"></i></button><pre class="downlit sourceCode r"><code class="sourceCode R"><span class="co"># Examine and visualize PCA results a few different ways</span>
<span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">[[</span><span class="st">"pca"</span><span class="op">]</span><span class="op">]</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, nfeatures <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></code></pre></div>
<pre><code>## PC_ 1 
## Positive:  CST3, TYROBP, LST1, AIF1, FTL 
## Negative:  MALAT1, LTB, IL32, IL7R, CD2 
## PC_ 2 
## Positive:  CD79A, MS4A1, TCL1A, HLA-DQA1, HLA-DQB1 
## Negative:  NKG7, PRF1, CST7, GZMB, GZMA 
## PC_ 3 
## Positive:  HLA-DQA1, CD79A, CD79B, HLA-DQB1, HLA-DPB1 
## Negative:  PPBP, PF4, SDPR, SPARC, GNG11 
## PC_ 4 
## Positive:  HLA-DQA1, CD79B, CD79A, MS4A1, HLA-DQB1 
## Negative:  VIM, IL7R, S100A6, IL32, S100A8 
## PC_ 5 
## Positive:  GZMB, NKG7, S100A8, FGFBP2, GNLY 
## Negative:  LTB, IL7R, CKB, VIM, MS4A7</code></pre>
<div class="sourceCode hasCopyButton" id="cb26"><button type="button" class="btn btn-primary btn-copy-ex" title="" aria-label="Copy to clipboard" data-toggle="tooltip" data-placement="left auto" data-trigger="hover" data-clipboard-copy="" data-original-title="Copy to clipboard"><i class="fa fa-copy"></i></button><pre class="downlit sourceCode r"><code class="sourceCode R"><span class="fu"><a href="https://satijalab.org/seurat/reference/VizDimLoadings.html">VizDimLoadings</a></span><span class="op">(</span><span class="va">pbmc</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">2</span>, reduction <span class="op">=</span> <span class="st">"pca"</span><span class="op">)</span></code></pre></div>
<p><img src="./Seurat - Guided Clustering Tutorial • Seurat_files/pca_viz-1.png" width="700"></p>
<div class="sourceCode hasCopyButton" id="cb27"><button type="button" class="btn btn-primary btn-copy-ex" title="" aria-label="Copy to clipboard" data-toggle="tooltip" data-placement="left auto" data-trigger="hover" data-clipboard-copy="" data-original-title="Copy to clipboard"><i class="fa fa-copy"></i></button><pre class="downlit sourceCode r"><code class="sourceCode R"><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">pbmc</span>, reduction <span class="op">=</span> <span class="st">"pca"</span><span class="op">)</span></code></pre></div>
<p><img src="./Seurat - Guided Clustering Tutorial • Seurat_files/pca_viz-2.png" width="700"></p>
<p>In particular <code><a href="https://satijalab.org/seurat/reference/DimHeatmap.html">DimHeatmap()</a></code> allows for easy exploration of the primary sources of heterogeneity in a dataset, and can be useful when trying to decide which PCs to include for further downstream analyses. Both cells and features are ordered according to their PCA scores. Setting <code>cells</code> to a number plots the ‘extreme’ cells on both ends of the spectrum, which dramatically speeds plotting for large datasets. Though clearly a supervised analysis, we find this to be a valuable tool for exploring correlated feature sets.</p>
<div class="sourceCode hasCopyButton" id="cb28"><button type="button" class="btn btn-primary btn-copy-ex" title="" aria-label="Copy to clipboard" data-toggle="tooltip" data-placement="left auto" data-trigger="hover" data-clipboard-copy="" data-original-title="Copy to clipboard"><i class="fa fa-copy"></i></button><pre class="downlit sourceCode r"><code class="sourceCode R"><span class="fu"><a href="https://satijalab.org/seurat/reference/DimHeatmap.html">DimHeatmap</a></span><span class="op">(</span><span class="va">pbmc</span>, dims <span class="op">=</span> <span class="fl">1</span>, cells <span class="op">=</span> <span class="fl">500</span>, balanced <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<p><img src="./Seurat - Guided Clustering Tutorial • Seurat_files/single-heatmap-1.png" width="700"></p>
<div class="sourceCode hasCopyButton" id="cb29"><button type="button" class="btn btn-primary btn-copy-ex" title="" aria-label="Copy to clipboard" data-toggle="tooltip" data-placement="left auto" data-trigger="hover" data-clipboard-copy="" data-original-title="Copy to clipboard"><i class="fa fa-copy"></i></button><pre class="downlit sourceCode r"><code class="sourceCode R"><span class="fu"><a href="https://satijalab.org/seurat/reference/DimHeatmap.html">DimHeatmap</a></span><span class="op">(</span><span class="va">pbmc</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">15</span>, cells <span class="op">=</span> <span class="fl">500</span>, balanced <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<p><img src="./Seurat - Guided Clustering Tutorial • Seurat_files/multi-heatmap-1.png" width="864"></p>
</div>
<div id="determine-the-dimensionality-of-the-dataset" class="section level1">
<h1 id="determine-the-dimensionality-of-the-dataset-1">
<a href="https://satijalab.org/seurat/articles/pbmc3k_tutorial.html#determine-the-dimensionality-of-the-dataset" class="anchor"></a>Determine the ‘dimensionality’ of the dataset</h1>
<p>To overcome the extensive technical noise in any single feature for scRNA-seq data, Seurat clusters cells based on their PCA scores, with each PC essentially representing a ‘metafeature’ that combines information across a correlated feature set. The top principal components therefore represent a robust compression of the dataset. However, how many components should we choose to include? 10? 20? 100?</p>
<p>In <a href="http://www.cell.com/abstract/S0092-8674(15)00549-8">Macosko <em>et al</em></a>, we implemented a resampling test inspired by the JackStraw procedure. We randomly permute a subset of the data (1% by default) and rerun PCA, constructing a ‘null distribution’ of feature scores, and repeat this procedure. We identify ‘significant’ PCs as those who have a strong enrichment of low p-value features.</p>
<div class="sourceCode hasCopyButton" id="cb30"><button type="button" class="btn btn-primary btn-copy-ex" title="" aria-label="Copy to clipboard" data-toggle="tooltip" data-placement="left auto" data-trigger="hover" data-clipboard-copy="" data-original-title="Copy to clipboard"><i class="fa fa-copy"></i></button><pre class="downlit sourceCode r"><code class="sourceCode R"><span class="co"># NOTE: This process can take a long time for big datasets, comment out for expediency. More</span>
<span class="co"># approximate techniques such as those implemented in ElbowPlot() can be used to reduce</span>
<span class="co"># computation time</span>
<span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/JackStraw.html">JackStraw</a></span><span class="op">(</span><span class="va">pbmc</span>, num.replicate <span class="op">=</span> <span class="fl">100</span><span class="op">)</span>
<span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/ScoreJackStraw.html">ScoreJackStraw</a></span><span class="op">(</span><span class="va">pbmc</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">20</span><span class="op">)</span></code></pre></div>
<p>The <code><a href="https://satijalab.org/seurat/reference/JackStrawPlot.html">JackStrawPlot()</a></code> function provides a visualization tool for comparing the distribution of p-values for each PC with a uniform distribution (dashed line). ‘Significant’ PCs will show a strong enrichment of features with low p-values (solid curve above the dashed line). In this case it appears that there is a sharp drop-off in significance after the first 10-12 PCs.</p>
<div class="sourceCode hasCopyButton" id="cb31"><button type="button" class="btn btn-primary btn-copy-ex" title="" aria-label="Copy to clipboard" data-toggle="tooltip" data-placement="left auto" data-trigger="hover" data-clipboard-copy="" data-original-title="Copy to clipboard"><i class="fa fa-copy"></i></button><pre class="downlit sourceCode r"><code class="sourceCode R"><span class="fu"><a href="https://satijalab.org/seurat/reference/JackStrawPlot.html">JackStrawPlot</a></span><span class="op">(</span><span class="va">pbmc</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">15</span><span class="op">)</span></code></pre></div>
<p><img src="./Seurat - Guided Clustering Tutorial • Seurat_files/jsplots-1.png" width="960"></p>
<p>An alternative heuristic method generates an ‘Elbow plot’: a ranking of principle components based on the percentage of variance explained by each one (<code><a href="https://satijalab.org/seurat/reference/ElbowPlot.html">ElbowPlot()</a></code> function). In this example, we can observe an ‘elbow’ around PC9-10, suggesting that the majority of true signal is captured in the first 10 PCs.</p>
<div class="sourceCode hasCopyButton" id="cb32"><button type="button" class="btn btn-primary btn-copy-ex" title="" aria-label="Copy to clipboard" data-toggle="tooltip" data-placement="left auto" data-trigger="hover" data-clipboard-copy="" data-original-title="Copy to clipboard"><i class="fa fa-copy"></i></button><pre class="downlit sourceCode r"><code class="sourceCode R"><span class="fu"><a href="https://satijalab.org/seurat/reference/ElbowPlot.html">ElbowPlot</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span></code></pre></div>
<p><img src="./Seurat - Guided Clustering Tutorial • Seurat_files/elbow_plot-1.png" width="960"></p>
<p>Identifying the true dimensionality of a dataset – can be challenging/uncertain for the user. We therefore suggest these three approaches to consider. The first is more supervised, exploring PCs to determine relevant sources of heterogeneity, and could be used in conjunction with GSEA for example. The second implements a statistical test based on a random null model, but is time-consuming for large datasets, and may not return a clear PC cutoff. The third is a heuristic that is commonly used, and can be calculated instantly. In this example, all three approaches yielded similar results, but we might have been justified in choosing anything between PC 7-12 as a cutoff.</p>
<p>We chose 10 here, but encourage users to consider the following:</p>
<ul>
<li>Dendritic cell and NK aficionados may recognize that genes strongly associated with PCs 12 and 13 define rare immune subsets (i.e.&nbsp;MZB1 is a marker for plasmacytoid DCs). However, these groups are so rare, they are difficult to distinguish from background noise for a dataset of this size without prior knowledge.</li>
<li>We encourage users to repeat downstream analyses with a different number of PCs (10, 15, or even 50!). As you will observe, the results often do not differ dramatically.</li>
<li>We advise users to err on the higher side when choosing this parameter. For example, performing downstream analyses with only 5 PCs does significantly and adversely affect results.</li>
</ul>
<hr>
</div>
<div id="cluster-the-cells" class="section level1">
<h1 id="cluster-the-cells-1">
<a href="https://satijalab.org/seurat/articles/pbmc3k_tutorial.html#cluster-the-cells" class="anchor"></a>Cluster the cells</h1>
<p>Seurat v3 applies a graph-based clustering approach, building upon initial strategies in (<a href="http://www.cell.com/abstract/S0092-8674(15)00549-8">Macosko <em>et al</em></a>). Importantly, the <em>distance metric</em> which drives the clustering analysis (based on previously identified PCs) remains the same. However, our approach to partitioning the cellular distance matrix into clusters has dramatically improved. Our approach was heavily inspired by recent manuscripts which applied graph-based clustering approaches to scRNA-seq data <a href="http://bioinformatics.oxfordjournals.org/content/early/2015/02/10/bioinformatics.btv088.abstract">[SNN-Cliq, Xu and Su, Bioinformatics, 2015]</a> and CyTOF data <a href="http://www.ncbi.nlm.nih.gov/pubmed/26095251">[PhenoGraph, Levine <em>et al</em>., Cell, 2015]</a>. Briefly, these methods embed cells in a graph structure - for example a K-nearest neighbor (KNN) graph, with edges drawn between cells with similar feature expression patterns, and then attempt to partition this graph into highly interconnected ‘quasi-cliques’ or ‘communities’.</p>
<p>As in PhenoGraph, we first construct a KNN graph based on the euclidean distance in PCA space, and refine the edge weights between any two cells based on the shared overlap in their local neighborhoods (Jaccard similarity). This step is performed using the <code><a href="https://satijalab.org/seurat/reference/FindNeighbors.html">FindNeighbors()</a></code> function, and takes as input the previously defined dimensionality of the dataset (first 10 PCs).</p>
<p>To cluster the cells, we next apply modularity optimization techniques such as the Louvain algorithm (default) or SLM <a href="http://dx.doi.org/10.1088/1742-5468/2008/10/P10008">[SLM, Blondel <em>et al</em>., Journal of Statistical Mechanics]</a>, to iteratively group cells together, with the goal of optimizing the standard modularity function. The <code><a href="https://satijalab.org/seurat/reference/FindClusters.html">FindClusters()</a></code> function implements this procedure, and contains a resolution parameter that sets the ‘granularity’ of the downstream clustering, with increased values leading to a greater number of clusters. We find that setting this parameter between 0.4-1.2 typically returns good results for single-cell datasets of around 3K cells. Optimal resolution often increases for larger datasets. The clusters can be found using the <code><a href="https://mojaveazure.github.io/seurat-object/reference/Idents.html">Idents()</a></code> function.</p>
<div class="sourceCode hasCopyButton" id="cb33"><button type="button" class="btn btn-primary btn-copy-ex" title="" aria-label="Copy to clipboard" data-toggle="tooltip" data-placement="left auto" data-trigger="hover" data-clipboard-copy="" data-original-title="Copy to clipboard"><i class="fa fa-copy"></i></button><pre class="downlit sourceCode r"><code class="sourceCode R"><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindNeighbors.html">FindNeighbors</a></span><span class="op">(</span><span class="va">pbmc</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span>
<span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindClusters.html">FindClusters</a></span><span class="op">(</span><span class="va">pbmc</span>, resolution <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></code></pre></div>
<pre><code>## Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck
## 
## Number of nodes: 2638
## Number of edges: 95927
## 
## Running Louvain algorithm...
## Maximum modularity in 10 random starts: 0.8728
## Number of communities: 9
## Elapsed time: 0 seconds</code></pre>
<div class="sourceCode hasCopyButton" id="cb35"><button type="button" class="btn btn-primary btn-copy-ex" title="" aria-label="Copy to clipboard" data-toggle="tooltip" data-placement="left auto" data-trigger="hover" data-clipboard-copy="" data-original-title="Copy to clipboard"><i class="fa fa-copy"></i></button><pre class="downlit sourceCode r"><code class="sourceCode R"><span class="co"># Look at cluster IDs of the first 5 cells</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/Idents.html">Idents</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span>, <span class="fl">5</span><span class="op">)</span></code></pre></div>
<pre><code>## AAACATACAACCAC-1 AAACATTGAGCTAC-1 AAACATTGATCAGC-1 AAACCGTGCTTCCG-1 
##                2                3                2                1 
## AAACCGTGTATGCG-1 
##                6 
## Levels: 0 1 2 3 4 5 6 7 8</code></pre>
<hr>
</div>
<div id="run-non-linear-dimensional-reduction-umaptsne" class="section level1">
<h1 id="run-non-linear-dimensional-reduction-umap-tsne-">
<a href="https://satijalab.org/seurat/articles/pbmc3k_tutorial.html#run-non-linear-dimensional-reduction-umaptsne" class="anchor"></a>Run non-linear dimensional reduction (UMAP/tSNE)</h1>
<p>Seurat offers several non-linear dimensional reduction techniques, such as tSNE and UMAP, to visualize and explore these datasets. The goal of these algorithms is to learn the underlying manifold of the data in order to place similar cells together in low-dimensional space. Cells within the graph-based clusters determined above should co-localize on these dimension reduction plots. As input to the UMAP and tSNE, we suggest using the same PCs as input to the clustering analysis.</p>
<div class="sourceCode hasCopyButton" id="cb37"><button type="button" class="btn btn-primary btn-copy-ex" title="" aria-label="Copy to clipboard" data-toggle="tooltip" data-placement="left auto" data-trigger="hover" data-clipboard-copy="" data-original-title="Copy to clipboard"><i class="fa fa-copy"></i></button><pre class="downlit sourceCode r"><code class="sourceCode R"><span class="co"># If you haven't installed UMAP, you can do so via reticulate::py_install(packages =</span>
<span class="co"># 'umap-learn')</span>
<span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html">RunUMAP</a></span><span class="op">(</span><span class="va">pbmc</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span></code></pre></div>
<div class="sourceCode hasCopyButton" id="cb38"><button type="button" class="btn btn-primary btn-copy-ex" title="" aria-label="Copy to clipboard" data-toggle="tooltip" data-placement="left auto" data-trigger="hover" data-clipboard-copy="" data-original-title="Copy to clipboard"><i class="fa fa-copy"></i></button><pre class="downlit sourceCode r"><code class="sourceCode R"><span class="co"># note that you can set `label = TRUE` or use the LabelClusters function to help label</span>
<span class="co"># individual clusters</span>
<span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">pbmc</span>, reduction <span class="op">=</span> <span class="st">"umap"</span><span class="op">)</span></code></pre></div>
<p><img src="./Seurat - Guided Clustering Tutorial • Seurat_files/tsneplot-1.png" width="672"></p>
<p>You can save the object at this point so that it can easily be loaded back in without having to rerun the computationally intensive steps performed above, or easily shared with collaborators.</p>
<div class="sourceCode hasCopyButton" id="cb39"><button type="button" class="btn btn-primary btn-copy-ex" title="" aria-label="Copy to clipboard" data-toggle="tooltip" data-placement="left auto" data-trigger="hover" data-clipboard-copy="" data-original-title="Copy to clipboard"><i class="fa fa-copy"></i></button><pre class="downlit sourceCode r"><code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">saveRDS</a></span><span class="op">(</span><span class="va">pbmc</span>, file <span class="op">=</span> <span class="st">"../output/pbmc_tutorial.rds"</span><span class="op">)</span></code></pre></div>
<hr>
</div>
<div id="finding-differentially-expressed-features-cluster-biomarkers" class="section level1">
<h1 id="finding-differentially-expressed-features-cluster-biomarkers-">
<a href="https://satijalab.org/seurat/articles/pbmc3k_tutorial.html#finding-differentially-expressed-features-cluster-biomarkers" class="anchor"></a>Finding differentially expressed features (cluster biomarkers)</h1>
<p>Seurat can help you find markers that define clusters via differential expression. By default, it identifies positive and negative markers of a single cluster (specified in <code>ident.1</code>), compared to all other cells. <code><a href="https://satijalab.org/seurat/reference/FindAllMarkers.html">FindAllMarkers()</a></code> automates this process for all clusters, but you can also test groups of clusters vs.&nbsp;each other, or against all cells.</p>
<p>The <code>min.pct</code> argument requires a feature to be detected at a minimum percentage in either of the two groups of cells, and the thresh.test argument requires a feature to be differentially expressed (on average) by some amount between the two groups. You can set both of these to 0, but with a dramatic increase in time - since this will test a large number of features that are unlikely to be highly discriminatory. As another option to speed up these computations, <code>max.cells.per.ident</code> can be set. This will downsample each identity class to have no more cells than whatever this is set to. While there is generally going to be a loss in power, the speed increases can be significant and the most highly differentially expressed features will likely still rise to the top.</p>
<div class="sourceCode hasCopyButton" id="cb40"><button type="button" class="btn btn-primary btn-copy-ex" title="" aria-label="Copy to clipboard" data-toggle="tooltip" data-placement="left auto" data-trigger="hover" data-clipboard-copy="" data-original-title="Copy to clipboard"><i class="fa fa-copy"></i></button><pre class="downlit sourceCode r"><code class="sourceCode R"><span class="co"># find all markers of cluster 2</span>
<span class="va">cluster2.markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindMarkers.html">FindMarkers</a></span><span class="op">(</span><span class="va">pbmc</span>, ident.1 <span class="op">=</span> <span class="fl">2</span>, min.pct <span class="op">=</span> <span class="fl">0.25</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">cluster2.markers</span>, n <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></code></pre></div>
<pre><code>##             p_val avg_log2FC pct.1 pct.2    p_val_adj
## IL32 2.892340e-90  1.2013522 0.947 0.465 3.966555e-86
## LTB  1.060121e-86  1.2695776 0.981 0.643 1.453850e-82
## CD3D 8.794641e-71  0.9389621 0.922 0.432 1.206097e-66
## IL7R 3.516098e-68  1.1873213 0.750 0.326 4.821977e-64
## LDHB 1.642480e-67  0.8969774 0.954 0.614 2.252497e-63</code></pre>
<div class="sourceCode hasCopyButton" id="cb42"><button type="button" class="btn btn-primary btn-copy-ex" title="" aria-label="Copy to clipboard" data-toggle="tooltip" data-placement="left auto" data-trigger="hover" data-clipboard-copy="" data-original-title="Copy to clipboard"><i class="fa fa-copy"></i></button><pre class="downlit sourceCode r"><code class="sourceCode R"><span class="co"># find all markers distinguishing cluster 5 from clusters 0 and 3</span>
<span class="va">cluster5.markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindMarkers.html">FindMarkers</a></span><span class="op">(</span><span class="va">pbmc</span>, ident.1 <span class="op">=</span> <span class="fl">5</span>, ident.2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">3</span><span class="op">)</span>, min.pct <span class="op">=</span> <span class="fl">0.25</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">cluster5.markers</span>, n <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></code></pre></div>
<pre><code>##                       p_val avg_log2FC pct.1 pct.2     p_val_adj
## FCGR3A        8.246578e-205   4.261495 0.975 0.040 1.130936e-200
## IFITM3        1.677613e-195   3.879339 0.975 0.049 2.300678e-191
## CFD           2.401156e-193   3.405492 0.938 0.038 3.292945e-189
## CD68          2.900384e-191   3.020484 0.926 0.035 3.977587e-187
## RP11-290F20.3 2.513244e-186   2.720057 0.840 0.017 3.446663e-182</code></pre>
<div class="sourceCode hasCopyButton" id="cb44"><button type="button" class="btn btn-primary btn-copy-ex" title="" aria-label="Copy to clipboard" data-toggle="tooltip" data-placement="left auto" data-trigger="hover" data-clipboard-copy="" data-original-title="Copy to clipboard"><i class="fa fa-copy"></i></button><pre class="downlit sourceCode r"><code class="sourceCode R"><span class="co"># find markers for every cluster compared to all remaining cells, report only the positive</span>
<span class="co"># ones</span>
<span class="va">pbmc.markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindAllMarkers.html">FindAllMarkers</a></span><span class="op">(</span><span class="va">pbmc</span>, only.pos <span class="op">=</span> <span class="cn">TRUE</span>, min.pct <span class="op">=</span> <span class="fl">0.25</span>, logfc.threshold <span class="op">=</span> <span class="fl">0.25</span><span class="op">)</span>
<span class="va">pbmc.markers</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">cluster</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/slice.html">slice_max</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">2</span>, order_by <span class="op">=</span> <span class="va">avg_log2FC</span><span class="op">)</span></code></pre></div>
<pre><code>## # A tibble: 18 × 7
## # Groups:   cluster [9]
##        p_val avg_log2FC pct.1 pct.2 p_val_adj cluster gene    
##        &lt;dbl&gt;      &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;     &lt;dbl&gt; &lt;fct&gt;   &lt;chr&gt;   
##  1 9.57e- 88       1.36 0.447 0.108 1.31e- 83 0       CCR7    
##  2 3.75e-112       1.09 0.912 0.592 5.14e-108 0       LDHB    
##  3 0               5.57 0.996 0.215 0         1       S100A9  
##  4 0               5.48 0.975 0.121 0         1       S100A8  
##  5 1.06e- 86       1.27 0.981 0.643 1.45e- 82 2       LTB     
##  6 2.97e- 58       1.23 0.42  0.111 4.07e- 54 2       AQP3    
##  7 0               4.31 0.936 0.041 0         3       CD79A   
##  8 9.48e-271       3.59 0.622 0.022 1.30e-266 3       TCL1A   
##  9 5.61e-202       3.10 0.983 0.234 7.70e-198 4       CCL5    
## 10 7.25e-165       3.00 0.577 0.055 9.95e-161 4       GZMK    
## 11 3.51e-184       3.31 0.975 0.134 4.82e-180 5       FCGR3A  
## 12 2.03e-125       3.09 1     0.315 2.78e-121 5       LST1    
## 13 3.13e-191       5.32 0.961 0.131 4.30e-187 6       GNLY    
## 14 7.95e-269       4.83 0.961 0.068 1.09e-264 6       GZMB    
## 15 1.48e-220       3.87 0.812 0.011 2.03e-216 7       FCER1A  
## 16 1.67e- 21       2.87 1     0.513 2.28e- 17 7       HLA-DPB1
## 17 1.92e-102       8.59 1     0.024 2.63e- 98 8       PPBP    
## 18 9.25e-186       7.29 1     0.011 1.27e-181 8       PF4</code></pre>
<p>Seurat has several tests for differential expression which can be set with the test.use parameter (see our <a href="https://satijalab.org/seurat/articles/de_vignette.html">DE vignette</a> for details). For example, the ROC test returns the ‘classification power’ for any individual marker (ranging from 0 - random, to 1 - perfect).</p>
<div class="sourceCode hasCopyButton" id="cb46"><button type="button" class="btn btn-primary btn-copy-ex" title="" aria-label="Copy to clipboard" data-toggle="tooltip" data-placement="left auto" data-trigger="hover" data-clipboard-copy="" data-original-title="Copy to clipboard"><i class="fa fa-copy"></i></button><pre class="downlit sourceCode r"><code class="sourceCode R"><span class="va">cluster0.markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindMarkers.html">FindMarkers</a></span><span class="op">(</span><span class="va">pbmc</span>, ident.1 <span class="op">=</span> <span class="fl">0</span>, logfc.threshold <span class="op">=</span> <span class="fl">0.25</span>, test.use <span class="op">=</span> <span class="st">"roc"</span>, only.pos <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<p>We include several tools for visualizing marker expression. <code><a href="https://satijalab.org/seurat/reference/VlnPlot.html">VlnPlot()</a></code> (shows expression probability distributions across clusters), and <code><a href="https://satijalab.org/seurat/reference/FeaturePlot.html">FeaturePlot()</a></code> (visualizes feature expression on a tSNE or PCA plot) are our most commonly used visualizations. We also suggest exploring <code><a href="https://satijalab.org/seurat/reference/RidgePlot.html">RidgePlot()</a></code>, <code><a href="https://satijalab.org/seurat/reference/CellScatter.html">CellScatter()</a></code>, and <code><a href="https://satijalab.org/seurat/reference/DotPlot.html">DotPlot()</a></code> as additional methods to view your dataset.</p>
<div class="sourceCode hasCopyButton" id="cb47"><button type="button" class="btn btn-primary btn-copy-ex" title="" aria-label="Copy to clipboard" data-toggle="tooltip" data-placement="left auto" data-trigger="hover" data-clipboard-copy="" data-original-title="Copy to clipboard"><i class="fa fa-copy"></i></button><pre class="downlit sourceCode r"><code class="sourceCode R"><span class="fu"><a href="https://satijalab.org/seurat/reference/VlnPlot.html">VlnPlot</a></span><span class="op">(</span><span class="va">pbmc</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"MS4A1"</span>, <span class="st">"CD79A"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="./Seurat - Guided Clustering Tutorial • Seurat_files/markerplots-1.png" width="1440"></p>
<div class="sourceCode hasCopyButton" id="cb48"><button type="button" class="btn btn-primary btn-copy-ex" title="" aria-label="Copy to clipboard" data-toggle="tooltip" data-placement="left auto" data-trigger="hover" data-clipboard-copy="" data-original-title="Copy to clipboard"><i class="fa fa-copy"></i></button><pre class="downlit sourceCode r"><code class="sourceCode R"><span class="co"># you can plot raw counts as well</span>
<span class="fu"><a href="https://satijalab.org/seurat/reference/VlnPlot.html">VlnPlot</a></span><span class="op">(</span><span class="va">pbmc</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"NKG7"</span>, <span class="st">"PF4"</span><span class="op">)</span>, slot <span class="op">=</span> <span class="st">"counts"</span>, log <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<p><img src="./Seurat - Guided Clustering Tutorial • Seurat_files/markerplots-2.png" width="1440"></p>
<div class="sourceCode hasCopyButton" id="cb49"><button type="button" class="btn btn-primary btn-copy-ex" title="" aria-label="Copy to clipboard" data-toggle="tooltip" data-placement="left auto" data-trigger="hover" data-clipboard-copy="" data-original-title="Copy to clipboard"><i class="fa fa-copy"></i></button><pre class="downlit sourceCode r"><code class="sourceCode R"><span class="fu"><a href="https://satijalab.org/seurat/reference/FeaturePlot.html">FeaturePlot</a></span><span class="op">(</span><span class="va">pbmc</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"MS4A1"</span>, <span class="st">"GNLY"</span>, <span class="st">"CD3E"</span>, <span class="st">"CD14"</span>, <span class="st">"FCER1A"</span>, <span class="st">"FCGR3A"</span>, <span class="st">"LYZ"</span>, <span class="st">"PPBP"</span>,
    <span class="st">"CD8A"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="./Seurat - Guided Clustering Tutorial • Seurat_files/markerplots-3.png" width="1440"></p>
<p><code><a href="https://satijalab.org/seurat/reference/DoHeatmap.html">DoHeatmap()</a></code> generates an expression heatmap for given cells and features. In this case, we are plotting the top 20 markers (or all markers if less than 20) for each cluster.</p>
<div class="sourceCode hasCopyButton" id="cb50"><button type="button" class="btn btn-primary btn-copy-ex" title="" aria-label="Copy to clipboard" data-toggle="tooltip" data-placement="left auto" data-trigger="hover" data-clipboard-copy="" data-original-title="Copy to clipboard"><i class="fa fa-copy"></i></button><pre class="downlit sourceCode r"><code class="sourceCode R"><span class="va">pbmc.markers</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">cluster</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/top_n.html">top_n</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">10</span>, wt <span class="op">=</span> <span class="va">avg_log2FC</span><span class="op">)</span> <span class="op">-&gt;</span> <span class="va">top10</span>
<span class="fu"><a href="https://satijalab.org/seurat/reference/DoHeatmap.html">DoHeatmap</a></span><span class="op">(</span><span class="va">pbmc</span>, features <span class="op">=</span> <span class="va">top10</span><span class="op">$</span><span class="va">gene</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html">NoLegend</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<p><img src="./Seurat - Guided Clustering Tutorial • Seurat_files/clusterHeatmap-1.png" width="1440"></p>
<hr>
</div>
<div id="assigning-cell-type-identity-to-clusters" class="section level1">
<h1 id="assigning-cell-type-identity-to-clusters-1">
<a href="https://satijalab.org/seurat/articles/pbmc3k_tutorial.html#assigning-cell-type-identity-to-clusters" class="anchor"></a>Assigning cell type identity to clusters</h1>
<p>Fortunately in the case of this dataset, we can use canonical markers to easily match the unbiased clustering to known cell types:</p>
<table class="table">
<thead><tr class="header">
<th>Cluster ID</th>
<th>Markers</th>
<th>Cell Type</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>0</td>
<td>IL7R, CCR7</td>
<td>Naive CD4+ T</td>
</tr>
<tr class="even">
<td>1</td>
<td>CD14, LYZ</td>
<td>CD14+ Mono</td>
</tr>
<tr class="odd">
<td>2</td>
<td>IL7R, S100A4</td>
<td>Memory CD4+</td>
</tr>
<tr class="even">
<td>3</td>
<td>MS4A1</td>
<td>B</td>
</tr>
<tr class="odd">
<td>4</td>
<td>CD8A</td>
<td>CD8+ T</td>
</tr>
<tr class="even">
<td>5</td>
<td>FCGR3A, MS4A7</td>
<td>FCGR3A+ Mono</td>
</tr>
<tr class="odd">
<td>6</td>
<td>GNLY, NKG7</td>
<td>NK</td>
</tr>
<tr class="even">
<td>7</td>
<td>FCER1A, CST3</td>
<td>DC</td>
</tr>
<tr class="odd">
<td>8</td>
<td>PPBP</td>
<td>Platelet</td>
</tr>
</tbody>
</table>
<div class="sourceCode hasCopyButton" id="cb51"><button type="button" class="btn btn-primary btn-copy-ex" title="" aria-label="Copy to clipboard" data-toggle="tooltip" data-placement="left auto" data-trigger="hover" data-clipboard-copy="" data-original-title="Copy to clipboard"><i class="fa fa-copy"></i></button><pre class="downlit sourceCode r"><code class="sourceCode R"><span class="va">new.cluster.ids</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Naive CD4 T"</span>, <span class="st">"CD14+ Mono"</span>, <span class="st">"Memory CD4 T"</span>, <span class="st">"B"</span>, <span class="st">"CD8 T"</span>, <span class="st">"FCGR3A+ Mono"</span>,
    <span class="st">"NK"</span>, <span class="st">"DC"</span>, <span class="st">"Platelet"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">new.cluster.ids</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/levels.html">levels</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span>
<span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/Idents.html">RenameIdents</a></span><span class="op">(</span><span class="va">pbmc</span>, <span class="va">new.cluster.ids</span><span class="op">)</span>
<span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">pbmc</span>, reduction <span class="op">=</span> <span class="st">"umap"</span>, label <span class="op">=</span> <span class="cn">TRUE</span>, pt.size <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html">NoLegend</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<p><img src="./Seurat - Guided Clustering Tutorial • Seurat_files/labelplot-1.png" width="864"></p>
<div class="sourceCode hasCopyButton" id="cb52"><button type="button" class="btn btn-primary btn-copy-ex" title="" aria-label="Copy to clipboard" data-toggle="tooltip" data-placement="left auto" data-trigger="hover" data-clipboard-copy="" data-original-title="Copy to clipboard"><i class="fa fa-copy"></i></button><pre class="downlit sourceCode r"><code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">saveRDS</a></span><span class="op">(</span><span class="va">pbmc</span>, file <span class="op">=</span> <span class="st">"../output/pbmc3k_final.rds"</span><span class="op">)</span></code></pre></div>
<details><p></p><summary><strong>Session Info</strong></summary><p></p>
<div class="sourceCode hasCopyButton" id="cb53"><button type="button" class="btn btn-primary btn-copy-ex" title="" aria-label="Copy to clipboard" data-toggle="tooltip" data-placement="left auto" data-trigger="hover" data-clipboard-copy="" data-original-title="Copy to clipboard"><i class="fa fa-copy"></i></button><pre class="downlit sourceCode r"><code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<pre><code>## R version 4.1.0 (2021-05-18)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Ubuntu 20.04.2 LTS
## 
## Matrix products: default
## BLAS/LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.8.so
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=C             
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] stats     graphics  grDevices utils     datasets  methods   base     
## 
## other attached packages:
## [1] ggplot2_3.3.5      patchwork_1.1.1    SeuratObject_4.0.4 Seurat_4.0.6      
## [5] dplyr_1.0.7       
## 
## loaded via a namespace (and not attached):
##   [1] Rtsne_0.15             colorspace_2.0-2       deldir_1.0-6          
##   [4] ellipsis_0.3.2         ggridges_0.5.3         rprojroot_2.0.2       
##   [7] fs_1.5.2               spatstat.data_2.1-2    farver_2.1.0          
##  [10] leiden_0.3.9           listenv_0.8.0          ggrepel_0.9.1         
##  [13] RSpectra_0.16-0        fansi_1.0.0            codetools_0.2-18      
##  [16] splines_4.1.0          cachem_1.0.6           knitr_1.33            
##  [19] polyclip_1.10-0        jsonlite_1.7.2         ica_1.0-2             
##  [22] cluster_2.1.2          png_0.1-7              uwot_0.1.11           
##  [25] shiny_1.7.1            sctransform_0.3.2.9011 spatstat.sparse_2.1-0 
##  [28] compiler_4.1.0         httr_1.4.2             assertthat_0.2.1      
##  [31] Matrix_1.3-3           fastmap_1.1.0          lazyeval_0.2.2        
##  [34] cli_3.1.0              limma_3.48.0           later_1.3.0           
##  [37] formatR_1.11           htmltools_0.5.2        tools_4.1.0           
##  [40] igraph_1.2.11          gtable_0.3.0           glue_1.6.0            
##  [43] RANN_2.6.1             reshape2_1.4.4         Rcpp_1.0.7            
##  [46] scattermore_0.7        jquerylib_0.1.4        pkgdown_1.6.1         
##  [49] vctrs_0.3.8            nlme_3.1-152           lmtest_0.9-39         
##  [52] xfun_0.25              stringr_1.4.0          globals_0.14.0        
##  [55] mime_0.12              miniUI_0.1.1.1         lifecycle_1.0.1       
##  [58] irlba_2.3.5            goftest_1.2-3          future_1.23.0         
##  [61] MASS_7.3-54            zoo_1.8-9              scales_1.1.1          
##  [64] spatstat.core_2.3-2    ragg_1.1.3             promises_1.2.0.1      
##  [67] spatstat.utils_2.3-0   parallel_4.1.0         RColorBrewer_1.1-2    
##  [70] yaml_2.2.1             memoise_2.0.0          reticulate_1.22       
##  [73] pbapply_1.5-0          gridExtra_2.3          sass_0.4.0            
##  [76] rpart_4.1-15           stringi_1.7.6          highr_0.9             
##  [79] desc_1.4.0             rlang_0.4.12           pkgconfig_2.0.3       
##  [82] systemfonts_1.0.2      matrixStats_0.61.0     evaluate_0.14         
##  [85] lattice_0.20-44        tensor_1.5             ROCR_1.0-11           
##  [88] purrr_0.3.4            labeling_0.4.2         htmlwidgets_1.5.4     
##  [91] cowplot_1.1.1          tidyselect_1.1.1       parallelly_1.30.0     
##  [94] RcppAnnoy_0.0.19       plyr_1.8.6             magrittr_2.0.1        
##  [97] R6_2.5.1               generics_0.1.1         DBI_1.1.1             
## [100] withr_2.4.3            mgcv_1.8-35            pillar_1.6.4          
## [103] fitdistrplus_1.1-6     abind_1.4-5            survival_3.2-11       
## [106] tibble_3.1.6           future.apply_1.8.1     crayon_1.4.2          
## [109] KernSmooth_2.23-20     utf8_1.2.2             spatstat.geom_2.3-1   
## [112] plotly_4.10.0          rmarkdown_2.10         grid_4.1.0            
## [115] data.table_1.14.2      digest_0.6.29          xtable_1.8-4          
## [118] tidyr_1.1.4            httpuv_1.6.5           textshaping_0.3.5     
## [121] munsell_0.5.0          viridisLite_0.4.0      bslib_0.3.1</code></pre>
</details>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip="">Contents</h2>
    <ul class="nav"><li><a href="https://satijalab.org/seurat/articles/pbmc3k_tutorial.html#setup-the-seurat-object-1">
Setup the Seurat Object</a></li><li><a href="https://satijalab.org/seurat/articles/pbmc3k_tutorial.html#standard-pre-processing-workflow-1">
Standard pre-processing workflow</a><ul class="nav"><li><a href="https://satijalab.org/seurat/articles/pbmc3k_tutorial.html#qc-and-selecting-cells-for-further-analysis-1">
QC and selecting cells for further analysis</a></li></ul></li><li><a href="https://satijalab.org/seurat/articles/pbmc3k_tutorial.html#normalizing-the-data-1">
Normalizing the data</a></li><li><a href="https://satijalab.org/seurat/articles/pbmc3k_tutorial.html#identification-of-highly-variable-features-feature-selection-">
Identification of highly variable features (feature selection)</a></li><li><a href="https://satijalab.org/seurat/articles/pbmc3k_tutorial.html#scaling-the-data-1">
Scaling the data</a></li><li><a href="https://satijalab.org/seurat/articles/pbmc3k_tutorial.html#perform-linear-dimensional-reduction-1">
Perform linear dimensional reduction</a></li><li><a href="https://satijalab.org/seurat/articles/pbmc3k_tutorial.html#determine-the-dimensionality-of-the-dataset-1">
Determine the ‘dimensionality’ of the dataset</a></li><li><a href="https://satijalab.org/seurat/articles/pbmc3k_tutorial.html#cluster-the-cells-1">
Cluster the cells</a></li><li><a href="https://satijalab.org/seurat/articles/pbmc3k_tutorial.html#run-non-linear-dimensional-reduction-umap-tsne-">
Run non-linear dimensional reduction (UMAP/tSNE)</a></li><li><a href="https://satijalab.org/seurat/articles/pbmc3k_tutorial.html#finding-differentially-expressed-features-cluster-biomarkers-">
Finding differentially expressed features (cluster biomarkers)</a></li><li><a href="https://satijalab.org/seurat/articles/pbmc3k_tutorial.html#assigning-cell-type-identity-to-clusters-1">
Assigning cell type identity to clusters</a></li></ul></nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Paul Hoffman, Satija Lab and Collaborators.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  

<script>mendeleyWebImporter = {
  downloadPdfs(e,t) { return this._call('downloadPdfs', [e,t]); },
  open() { return this._call('open', []); },
  setLoginToken(e) { return this._call('setLoginToken', [e]); },
  _call(methodName, methodArgs) {
    const id = Math.random();
    window.postMessage({ id, token: '0.5656682966714284', methodName, methodArgs }, 'https://satijalab.org');
    return new Promise(resolve => {
      const listener = window.addEventListener('message', event => {
        const data = event.data;
        if (typeof data !== 'object' || !('result' in data) || data.id !== id) return;
        window.removeEventListener('message', listener);
        resolve(data.result);
      });
    });
  }
};</script></body></html>